ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -122,119 +123,223 @@ hexDump :: BL.ByteString -> String -> IO ()
-#endif
-                Socket.bind sd (Socket.addrAddress nodeAddress)
-            Socket.connect sd (Socket.addrAddress remoteAddr)
-            bearer <- socketAsMuxBearer sd
-            Mx.muxBearerSetState bearer Mx.Connected
-            kConn $ Connection {
-                runConnection =
-                  Mx.muxStart mpds bearer
-                  `catch`
-                  handleMuxError
-              }
-        )
-
-    -- catch @'MuxBearerClosed'@ exception; we should ignore it and let @kConn@
-    -- finish; @connect@ will close the underlying socket.
-    --
-    -- Note: we do it only for initiated connections, not ones that the server
-    -- accepted.  The assymetry comes simply from the fact that in initiated
-    -- connections we might want to run a computation that is not interrupted
-    -- by a normal shutdown (a received terminal message throws
-    -- @'Mx.MuxBearerClosed'@ exception).
-    handleMuxError :: Mx.MuxError -> IO ()
-    handleMuxError Mx.MuxError { Mx.errorType = Mx.MuxBearerClosed } = return ()
-    handleMuxError e                                                 = throwIO e
+    case addr of
+      Nothing -> pure ()
+      Just addr_ -> do
+        Socket.bind sd addr_
+        Socket.listen sd 1
+    pure sd
+
+
+-- |
+-- Make a server-compatible socket from a network socket.
+--
+fromSocket
+    :: Socket.Socket
+    -> Server.Socket Socket.SockAddr Socket.Socket
+fromSocket sd = Server.Socket
+    { Server.acceptConnection = do
+        (sd', addr) <- Socket.accept sd
+        pure (addr, sd', Socket.close sd')
+    }
+
+
+-- |
+-- Thin wrapper around @'Server.run'@.
+--
+runNetworkNode'
+    :: forall ptcl st t.
+       ( Mx.ProtocolEnum ptcl
+       , Ord ptcl
+       , Enum ptcl
+       , Bounded ptcl
+       , Show ptcl
+       , Mx.MiniProtocolLimits ptcl
+       )
+    => Socket.Socket
+    -> (SomeException -> IO ())
+    -> (Socket.SockAddr -> st -> STM.STM (AcceptConnection st ptcl IO))
+    -> Server.CompleteConnection st ()
+    -> Server.Main st t
+    -> st
+    -> IO t
+runNetworkNode' sd acceptException acceptConn complete main st =
+    Server.run (fromSocket sd) acceptException (beginConnection acceptConn) complete main st
+
+
+-- |
+-- Run a server application.  It will listen on the given address for incoming
+-- connection.  The server thread runs using @withAsync@ function, which means
+-- that it will terminate when the callback terminates or throws an exception.
+--
+-- TODO: we should track connections in the state and refuse connections from
+-- peers we are already connected to.  This is also the right place to ban
+-- connection from peers which missbehaved.
+withServerNode
+    :: forall ptcl t.
+       ( Mx.ProtocolEnum ptcl
+       , Ord ptcl
+       , Enum ptcl
+       , Bounded ptcl
+       , Show ptcl
+       , Mx.MiniProtocolLimits ptcl
+       )
+    => Socket.AddrInfo
+    -> (Socket.SockAddr -> AnyMuxResponderApp ptcl IO)
```

Comment:
Is it intentional that we can change the definition of the responder application according to the address of the initiator? Seems like a mistake.

Reply:
This probably doesn't make sense without access to the state like in `beginConnection`; I'll simplify it here.